docs/day2/02_fastqc.md

```diff
@@ -0,0 +1,329 @@
+# 2.2 Samplesheets, operators, and groovy  
+
+!!! note "Learning objectives"
+
+    1. Implement a process with a tuple input
+    2. Understand why samplesheets should be used to read in data
+    3. Create a channel with operators and Groovy
+
+In this step we will implement `01_fastqc.sh` as a process called `FASTQC` and
+use operators and Groovy to create a channel.  
+
+![](img/2.excalidraw.png)
+
+```bash title="01_fastqc.sh"
+SAMPLE_ID=gut
+READS_1="data/ggal/${SAMPLE_ID}_1.fq"
+READS_2="data/ggal/${SAMPLE_ID}_2.fq"
+
+mkdir -p "results/fastqc_${SAMPLE_ID}_logs"
+fastqc --outdir "results/fastqc_${SAMPLE_ID}_logs" --format fastq ${READS_1} ${READS_2}
+```
+
+There's a lot going on in this script, let's break it down.
+
+`SAMPLE_ID=gut` assigns "gut" to the bash variable`SAMPLE_ID`. This is used to:  
+
+- Avoid hardcoding the sample name multiple times in the script  
+- Ensure that file pairs of the same sample are processed together  
+- Ensure that this script can be run on different sample pairs  
+
+`READS_1` and `READS_2` specify the paths to the `.fq` files.  
+
+Similar to the bash script in the previous step (`00_index.sh`), `mkdir -p`
+creates an output folder so that the `fastqc` outputs can be saved here.  
+
+In the `fastqc` command,
+
+- `--outdir` specifies the name of the output directory
+- `--format` is a required flag to indicate what format the the reads are in
+- `${READS_1}` and `${READS_2}` propagate the paths of the `.fq` files  
+
+## 2.2.1 Building the process  
+
+Start by adding the following `process` scaffold and script definition:  
+
+```groovy title="main.nf"
+process FASTQC {
+  [ directives ]
+
+  input:
+    < process inputs >
+
+  output:
+    < process outputs >
+
+  script:
+  """
+  mkdir -p "fastqc_${sample_id}_logs"
+  fastqc --outdir "fastqc_${sample_id}_logs" --format fastq $reads_1 $reads_2
+  """
+}
+```
+
+> Note about ${}  
+
+Unlike `salmon` from the previous process, `fastqc` requires that the output
+directory be created before running the command, hence the `mkdir ...` line.
+
+!!! question "Exercise"
+
+    Add the output definition
+
+    ??? note "Solution"
+
+        ```groovy title="main.nf"
+        output:
+        path "fastqc_${sample_id}_logs"
+        ```
+
+There are three inputs for this process definition that can be taken from the
+script definition you just added:
+
+1. `$sample_id`
+2. `$reads_1`
+3. `$reads_2`
+
+We will use a tuple as the input qualifier as it's useful to group related
+inputs, or, inputs that need to be processed together such as in this case.  
```

Comment:
```suggestion
We need to group these inputs together so they can be processed as a single unit. This is a requirement when working with multiple pieces of data that are specific to a sample. 

!!! note "Importance of proper data grouping when using Nextflow"
  
  Nextflow uses channels to run processes in parallel and if you aren't careful about how handle multiple pieces of data that need to be tied together, you may mix datasets up. 
  
   We can use the [input qualifier](https://www.nextflow.io/docs/latest/process.html#inputs) `tuple` to group multiple values into a single input definition. 

In this process, we will use a tuple as the input qualifier because it allows us to group related inputs together. This is useful when we need to process several pieces of data at the same time, like in this case where we have `$sample_id` and its two read files `$reads_1` and `$reads_2`. 

The tuple ensures that these inputs stay linked and are processed as a set, preventing them from getting mixed up with other data.
```